### PR TITLE
feat: NodeExit future core implementation

### PR DESCRIPTION
**Pull request by allnil**
#6098

Prepared a future. Want to take a moment to think how to move the one-shot channel and "terminate" flag from the handle to the future in a nice way.

### CODE REVIEW

File: bin/reth/src/builder/mod.rs [hunk 1]
```diff
@@ -1356,28 +1360,47 @@ pub struct NodeHandle {
...
+    pub terminate: bool,
+}
+
+impl Future for NodeExitFuture {
+    type Output = eyre::Result<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let pin = self.get_mut();
+
+        match pin.consensus_engine_rx.poll_unpin(cx) {
```
Comment (mattsse) [inline bin/reth/src/builder/mod.rs:1393]: this needs to handle the case where the oneshot already returned Ready, and should be be polled again, relevant, see poll impl of https://docs.rs/futures/latest/futures/future/struct.Fuse.html

File: bin/reth/src/builder/mod.rs [hunk 2]
```diff
@@ -1356,28 +1360,47 @@ pub struct NodeHandle {
...
         &self.rpc_server_handles
     }
+}
 
-    /// Waits for the node to exit, if it was configured to exit.
-    pub async fn wait_for_node_exit(self) -> eyre::Result<()> {
-        self.consensus_engine_rx.await??;
+/// A Future which resolves when the node exits
+#[derive(Debug)]
+pub struct NodeExitFuture {
```
Comment (mattsse) [inline bin/reth/src/builder/mod.rs:1378]: I think for now we can move this inside the NodeHandle, replacing the existing fields

**Review by mattsse — CHANGES_REQUESTED** [review]

File: bin/reth/src/commands/node/mod.rs [hunk 3]
```diff
@@ -217,7 +217,11 @@ impl<Ext: RethCliExt> NodeCommand<Ext> {
         let handle = node_config.launch::<Ext>(ext, executor).await?;
 
         // wait for node exit
-        handle.wait_for_node_exit().await
+        let node_exit_future = NodeExitFuture {
```
Comment (gakonst) [inline bin/reth/src/commands/node/mod.rs:220]: @mattsse this seems worse / more verbose than before, why did we want this change?

Comment (mattsse) [reply bin/reth/src/commands/node/mod.rs:220]: see comment above, this should replace the existing fields

File: bin/reth/src/commands/node/mod.rs [hunk 4]
```diff
@@ -215,9 +215,7 @@ impl<Ext: RethCliExt> NodeCommand<Ext> {
 
         // launch the node
         let handle = node_config.launch::<Ext>(ext, executor).await?;
-
-        // wait for node exit
-        handle.wait_for_node_exit().await
```
Comment (mattsse) [inline bin/reth/src/commands/node/mod.rs:220]: I'd like to keep this function

File: bin/reth/src/builder/mod.rs [hunk 5]
```diff
@@ -1357,30 +1360,59 @@ pub struct NodeHandle {
...
 
-        if self.terminate {
-            Ok(())
+impl Future for NodeExitFuture {
+    type Output = eyre::Result<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        if let Some(rx) = this.consensus_engine_rx.as_mut() {
+            match rx.poll_unpin(cx) {
```
Comment (mattsse) [inline bin/reth/src/builder/mod.rs:1401]: this can be `match ready!()` I believe

**Review by mattsse — CHANGES_REQUESTED** [review]
nice,

this should be easily testable,
so let's add a few sanity tests for different variants

**Review by Rjected — APPROVED** [review]
this looks good to me, pending @mattsse

File: crates/node-core/src/node_config.rs [hunk 6]
```diff
@@ -1452,6 +1488,36 @@ mod tests {
...
+
+        let _ = tx.send(Ok(()));
+
+        let node_exit_future = NodeExitFuture::new(rx, false);
+        match timeout(Duration::from_millis(100), node_exit_future).await {
+            Ok(_) => panic!("terminations flag was set to FALSE - future shouldn't be resolved"),
+            Err(_) => {
+                println!("termination flag was set to FALSE - future expectadly didn't resolve")
+            }
+        }
```
Comment (mattsse) [inline crates/node-core/src/node_config.rs:1518]: pedantic nit:

this check can be done via `poll_fn(|cx| node_exit_future.poll(x))` so we don't need the timeout here

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]
nice!

**Pull request merged**